connection_producer.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package snowflake
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"database/sql"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"regexp"
+	"sync"
+
+	"github.com/hashicorp/vault/sdk/database/helper/connutil"
+	"github.com/mitchellh/mapstructure"
+	"github.com/snowflakedb/gosnowflake"
+)
+
+var (
+	ErrInvalidSnowflakeURL           = fmt.Errorf("invalid connection URL format, expect <account_name>.snowflakecomputing.com/<db_name>")
+	accountAndDBNameFromConnURLRegex = regexp.MustCompile(`^(.+)\.snowflakecomputing.com/(.+)$`) // Expected format: <account_name>.snowflakecomputing.com/<db_name>
+)
+
+type snowflakeConnectionProducer struct {
+	ConnectionURL         string `json:"connection_url"`
+	MaxOpenConnections    int    `json:"max_open_connections"`
+	MaxIdleConnections    int    `json:"max_idle_connections"`
+	MaxConnectionLifetime string `json:"max_connection_lifetime"`
+	Username              string `json:"username"`
+	Password              string `json:"password"`
+	PrivateKey            string `json:"private_key"`
+	UsernameTemplate      string `json:"username_template"`
+	DisableEscaping       bool   `json:"disable_escaping"`
+
+	Initialized bool
+	RawConfig   map[string]any
+	Type        string
+	snowflakeDB *sql.DB
+	sync.Mutex
```

Comment:
Looks like we have two locks. This one and the one in the `SnowflakeSQL` struct in snowflake.go. It looks like we have had 2 for a while since #20 incorrectly introduced it. I think we should remove the lock in the `SnowflakeSQL` in favor of this one. To be clear, we should still lock in NewUser, etc., but we should use the same lock as the one used on `Init`.

Also, can we give this a name? This is a bit of a matter of taste. However, if we embed it like this without a name then language servers in code editors can't find references to it.